src/jit/codegenxarch.cpp

```diff
@@ -2689,84 +2632,97 @@ CodeGen::genStoreRegisterReturnInLclVar(GenTreePtr treeNode)
         regNumber retReg0 = REG_NA;
         regNumber retReg1 = REG_NA;
 
-        emitAttr size0 = EA_UNKNOWN;
-        emitAttr size1 = EA_UNKNOWN;
-
         unsigned __int8 offset0 = 0;
         unsigned __int8 offset1 = 0;
 
         var_types type0 = TYP_UNKNOWN;
         var_types type1 = TYP_UNKNOWN;
 
-        bool firstIntUsed = false;
-        bool firstFloatUsed = false;
+        getStructTypeOffset(structDesc, &type0, &type1, &offset0, &offset1);
+        getStructReturnRegisters(type0, type1, &retReg0, &retReg1);
+
+        assert(retReg0 != REG_NA && retReg1 != REG_NA);
 
-        genGetStructTypeSizeOffset(structDesc, &type0, &type1, &size0, &size1, &offset0, &offset1);
+        getEmitter()->emitIns_S_R(ins_Store(type0), emitTypeSize(type0), retReg0, lclVarPtr->gtLclNum, offset0);
+        getEmitter()->emitIns_S_R(ins_Store(type1), emitTypeSize(type1), retReg1, lclVarPtr->gtLclNum, offset1);
 
-        if (type0 != TYP_UNKNOWN)
+        return true;
+    }
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+    return false;
+}
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+//------------------------------------------------------------------------
+// getStructReturnRegisters: Returns the return registers for a specific struct types.
+//
+// Arguments:
+//    type0         - the type of the first eightbyte to be returned.
+//    type1         - the type of the second eightbyte to be returned.
+//    retRegPtr0    - returns the register for the first eightbyte.
+//    retRegPtr1    - returns the register for the second eightbyte.
+//
+
+void
+CodeGen::getStructReturnRegisters(var_types type0,
+                                        var_types type1,
+                                        regNumber* retRegPtr0,
+                                        regNumber* retRegPtr1)
+{
+    *retRegPtr0 = REG_NA;
+    *retRegPtr1 = REG_NA;
+
+    bool firstIntUsed = false;
+    bool firstFloatUsed = false;
+
+    if (type0 != TYP_UNKNOWN)
+    {
+        if (varTypeIsI(type0) || varTypeIsIntegral(type0))
         {
-            if (structDesc.eightByteClassifications[0] == SystemVClassificationTypeIntegerReference ||
-                structDesc.eightByteClassifications[0] == SystemVClassificationTypeInteger)
-            {
-                retReg0 = REG_INTRET;
-                firstIntUsed = true;
-            }
-            else if (structDesc.eightByteClassifications[0] == SystemVClassificationTypeSSE)
+            *retRegPtr0 = REG_INTRET;
+            firstIntUsed = true;
+        }
+        else if (varTypeIsFloating(type0))
+        {
+            *retRegPtr0 = REG_FLOATRET;
+            firstFloatUsed = true;
+        }
+        else
+        {
+            assert(false && "Invalid eightbyte type");
+        }
+    }
+
+    if (type1 != TYP_UNKNOWN)
+    {
+        if (varTypeIsI(type1) || varTypeIsIntegral(type1))
```

Comment:
The above if-condition could be replaced with varTypeIsIntegralOrI(type0)
 Add a line note